### PR TITLE
FIX Check if future is done before trying to set a result on it

### PR DESCRIPTION
**Pull request by hoodmane**
Otherwise, future raises invalid state error. Question: should we try to avoid getting to the point where we call this in the first place??

cc @CNSeniorious000 this fixes the invalid state errors on https://github.com/pyodide/micropip/pull/111

### Description

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [x] Add / update tests

### CODE REVIEW

**Comment by CNSeniorious000** [comment]
So where we called this

**Comment by hoodmane** [comment]
It's a bit complicated, I haven't sifted through the full control flow. Here's the full traceback, but it's hard to find anything micropip related there. Presumably we cancelled a Future and then it later got sent a result from the browser event loop. We need to throw it away if we're already done.

<details>
<summary>Traceback</summary>

```
$PyObject_CallOneArg @ VM137 pyodide.asm.wasm-02a1db6a:1
$create_promise_handles_result_helper @ VM137 pyodide.asm.wasm-02a1db6a:1
onFulfilled @ VM129 pyodide.asm.js:1938
processTicksAndRejections @ node:internal/process/task_queues:95
Promise.then (async)
JsvObject_CallMethod @ VM129 pyodide.asm.js:2369
$JsvObject_CallMethodId @ VM137 pyodide.asm.wasm-02a1db6a:1
$wrap_promise @ VM137 pyodide.asm.wasm-02a1db6a:1
$Js2Py_func_default_call_result @ VM137 pyodide.asm.wasm-02a1db6a:1
$Js2PyConverter_convert @ VM137 pyodide.asm.wasm-02a1db6a:1
$JsMethod_Vectorcall_impl @ VM137 pyodide.asm.wasm-02a1db6a:1
$JsMethod_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_EvalFrameDefault @ VM137 pyodide.asm.wasm-02a1db6a:1
$gen_send_ex2 @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyGen_am_send @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyIter_Send @ VM137 pyodide.asm.wasm-02a1db6a:1
$task_step_impl @ VM137 pyodide.asm.wasm-02a1db6a:1
$task_step @ VM137 pyodide.asm.wasm-02a1db6a:1
$TaskStepMethWrapper_call @ VM137 pyodide.asm.wasm-02a1db6a:1
_PyEM_TrampolineCall_JS @ VM129 pyodide.asm.js:7851
$_PyObject_MakeTpCall @ VM137 pyodide.asm.wasm-02a1db6a:1
$context_run @ VM137 pyodide.asm.wasm-02a1db6a:1
$cfunction_vectorcall_FASTCALL_KEYWORDS @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyVectorcall_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyObject_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_EvalFrameDefault @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_Vector @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyFunction_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$_pyproxy_apply @ VM137 pyodide.asm.wasm-02a1db6a:1
callPyObjectKwargs @ VM129 pyodide.asm.js:4265
Module.callPyObjectMaybePromising @ VM129 pyodide.asm.js:4325
wrapper @ VM129 pyodide.asm.js:1859
processImmediate @ node:internal/timers:478
callbackTrampoline @ node:internal/async_hooks:130
Immediate (async)
init @ node:internal/inspector_async_hook:25
emitInitNative @ node:internal/async_hooks:202
emitInitScript @ node:internal/async_hooks:505
initAsyncResource @ node:internal/timers:160
Immediate @ node:internal/timers:624
setImmediate @ node:timers:306
scheduleCallbackImmediate @ VM129 pyodide.asm.js:6443
scheduleCallback @ VM129 pyodide.asm.js:6453
JsvFunction_CallBound @ VM129 pyodide.asm.js:2447
$JsMethod_Vectorcall_impl @ VM137 pyodide.asm.wasm-02a1db6a:1
$JsMethod_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_EvalFrameDefault @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_Vector @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyFunction_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$method_vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyVectorcall_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyObject_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_EvalFrameDefault @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_Vector @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyFunction_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_VectorcallMethod @ VM137 pyodide.asm.wasm-02a1db6a:1
$task_call_step_soon @ VM137 pyodide.asm.wasm-02a1db6a:1
$_asyncio_Task___init__ @ VM137 pyodide.asm.wasm-02a1db6a:1
$type_call @ VM137 pyodide.asm.wasm-02a1db6a:1
_PyEM_TrampolineCall_JS @ VM129 pyodide.asm.js:7851
$_PyObject_MakeTpCall @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_EvalFrameDefault @ VM137 pyodide.asm.wasm-02a1db6a:1
$gen_send_ex2 @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyGen_am_send @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyIter_Send @ VM137 pyodide.asm.wasm-02a1db6a:1
$task_step_impl @ VM137 pyodide.asm.wasm-02a1db6a:1
$task_step @ VM137 pyodide.asm.wasm-02a1db6a:1
$task_wakeup @ VM137 pyodide.asm.wasm-02a1db6a:1
_PyEM_TrampolineCall_JS @ VM129 pyodide.asm.js:7851
$cfunction_vectorcall_O @ VM137 pyodide.asm.wasm-02a1db6a:1
$context_run @ VM137 pyodide.asm.wasm-02a1db6a:1
$cfunction_vectorcall_FASTCALL_KEYWORDS @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyVectorcall_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyObject_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_EvalFrameDefault @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_Vector @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyFunction_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$_pyproxy_apply @ VM137 pyodide.asm.wasm-02a1db6a:1
callPyObjectKwargs @ VM129 pyodide.asm.js:4265
Module.callPyObjectMaybePromising @ VM129 pyodide.asm.js:4325
wrapper @ VM129 pyodide.asm.js:1859
processImmediate @ node:internal/timers:478
callbackTrampoline @ node:internal/async_hooks:130
Immediate (async)
init @ node:internal/inspector_async_hook:25
emitInitNative @ node:internal/async_hooks:202
emitInitScript @ node:internal/async_hooks:505
initAsyncResource @ node:internal/timers:160
Immediate @ node:internal/timers:624
setImmediate @ node:timers:306
scheduleCallbackImmediate @ VM129 pyodide.asm.js:6443
scheduleCallback @ VM129 pyodide.asm.js:6453
JsvFunction_CallBound @ VM129 pyodide.asm.js:2447
$JsMethod_Vectorcall_impl @ VM137 pyodide.asm.wasm-02a1db6a:1
$JsMethod_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_EvalFrameDefault @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_Vector @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyFunction_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$method_vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyVectorcall_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyObject_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Call @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_EvalFrameDefault @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_Vector @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyFunction_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_VectorcallMethod @ VM137 pyodide.asm.wasm-02a1db6a:1
$future_schedule_callbacks @ VM137 pyodide.asm.wasm-02a1db6a:1
$_asyncio_Future_set_result @ VM137 pyodide.asm.wasm-02a1db6a:1
$method_vectorcall_FASTCALL_KEYWORDS_METHOD @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_EvalFrameDefault @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyEval_Vector @ VM137 pyodide.asm.wasm-02a1db6a:1
$_PyFunction_Vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$method_vectorcall @ VM137 pyodide.asm.wasm-02a1db6a:1
$PyObject_CallOneArg @ VM137 pyodide.asm.wasm-02a1db6a:1
$create_promise_handles_result_helper @ VM137 pyodide.asm.wasm-02a1db6a:1
onFulfilled @ VM129 pyodide.asm.js:1938
processTicksAndRejections @ node:internal/process/task_queues:95
```
</details>

**Comment by CNSeniorious000** [comment]
> Presumably we cancelled a Future and then it later got sent a result from the browser event loop.

It's probably `fetch`, because it is native JavaScript async function which don't check the cancel state when scheduled.

```js
from js import fetch

async def f():
    while True:
        await fetch("/")

from asyncio import *

f = ensure_future(f())

f.cancel()
```

Running this and you will see a similar `asyncio.exceptions.InvalidStateError: invalid state` in the devtools console

**Comment by hoodmane** [comment]
Thanks for the test case.

File: src/core/jsproxy.c [hunk 1]
```diff
@@ -2648,9 +2650,15 @@ wrap_promise(JsVal promise, JsVal done_callback, PyObject* js2py_converter)
   result = _PyObject_CallMethodIdNoArgs(loop, &PyId_create_future);
   FAIL_IF_NULL(result);
 
-  set_result = _PyObject_GetAttrId(result, &PyId_set_result);
+  helpers_mod = PyImport_ImportModule("_pyodide._future_helpers");
+  FAIL_IF_NULL(helpers_mod);
+  _Py_IDENTIFIER(get_future_resolvers);
+  helpers =
+    _PyObject_CallMethodIdOneArg(helpers, &PyId_get_future_resolvers, result);
```
Comment (ryanking13) [inline src/core/jsproxy.c:2657]: I guess this needs to be helpers_mod not helpers?

```suggestion
    _PyObject_CallMethodIdOneArg(helpers_mod, &PyId_get_future_resolvers, result);
```

Comment (hoodmane) [reply src/core/jsproxy.c]: thanks!

**Pull request merged**